### PR TITLE
Implement recursive mkdir using std::filesystem in utils

### PR DESCRIPTION
**Pull request by tbhaxor**
## PR Summary

User can provide parts of the directories that doesn't exist. The earlier version just created top level directories and failed to recursively create parent-directories.

This PR check for the directory path and create recursively.

## PR Checklist

- [x] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/utils/change_rundir.cpp
Comment (Yurlungur) [inline src/utils/change_rundir.cpp]:
```suggestion
    msg << "### FATAL ERROR in function [ChangeToRunDir]" << std::endl
        << "Cannot create directory '" << pdir << "";
```

File: src/utils/change_rundir.cpp
Comment (Yurlungur) [inline src/utils/change_rundir.cpp]: I know you copied the code below, but I think we want to throw an exception here if we can.
```suggestion
      PARTHENON_THROW(msg);
```

File: src/utils/change_rundir.cpp
Comment (Yurlungur) [inline src/utils/change_rundir.cpp]: The linter will likely complain about the includes here not being in in order. You should stick this include one block up with the other C++ includes, in alphabetical order.

Once I trigger the workflow the tests, a format and linter check will be run. You can make sure they're correct by building the code and typing `make format` or `make lint` for the formatter and linter. `make format` will just format the code and you can commit after calling it. `make lint` will check that the linter likes everything and it will make suggestions.

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Thanks for the contribution! This seems like a nice little improvement. Can you add the PR to `CHANGELOG.md`? Pending that, tests and formatting passing, and a few  nitpicks below, I'm happy to approve.

As an aside, how did you come upon Parthenon? I'm always curious how people reach us.

**Comment by tbhaxor** [comment]
Hey @Yurlungur, could you re-review my PR?

**Comment by tbhaxor** [comment]
> As an aside, how did you come upon Parthenon? I'm always curious how people reach us.

I am working athenapk

**Comment by Yurlungur** [comment]
Looks like you're still missing the CHANGELOG. Did you forget to commit it? I'm going to trigger the automated tests now.

**Comment by tbhaxor** [comment]
> Looks like you're still missing the CHANGELOG. Did you forget to commit it? I'm going to trigger the automated tests now.

Yeah my bad, forgot to commit. Just did and pushed

**Review by Yurlungur — APPROVED** [review]
Assuming automated tests pass, I approve. However, our process requires two maintainers to review each PR. I've tagged @pgrete for a second quick review.

**Comment by tbhaxor** [comment]
Oh hi @pgrete. I did't know you maintain parthenon as well. Nice to meet you here 😄

**Comment by tbhaxor** [comment]
@Yurlungur I can't see these 5 checks output, are you running on self hosted containers or outside github actions?

File: src/utils/change_rundir.cpp
Comment (pgrete) [inline src/utils/change_rundir.cpp:46]: This create the directory with permissions 777, doesn't it (see https://en.cppreference.com/w/cpp/filesystem/create_directory "Creates the directory p as if by POSIX [mkdir()](http://pubs.opengroup.org/onlinepubs/9699919799/functions/mkdir.html) with a second argument of static_cast<int>([std::filesystem::perms::all](http://en.cppreference.com/w/cpp/filesystem/perms))"?
That's a little bit too permissive from my point of view and I'd feel more happy with 775 as before.

File: CHANGELOG.md
Comment (pgrete) [inline CHANGELOG.md]: I'd move this to the "Infrastructure" category as this is not really changing anything exposed.

**Comment by pgrete** [comment]
> @Yurlungur I can't see these 5 checks output, are you running on self hosted containers or outside github actions?

Those 5 tests are the "extended" tests. They take more time and are only triggered once we start to final auto-merging.

File: src/utils/change_rundir.cpp
Comment (tbhaxor) [reply src/utils/change_rundir.cpp:46]: Yes we can do next call fs::permissions to do that. I will do it

Comment (tbhaxor) [reply src/utils/change_rundir.cpp:46]: Lets use a bit more secure 0755 permission. This will allow group users to read and cd into this directory but not write into it. Done in 97d088a4

File: CHANGELOG.md
Comment (tbhaxor) [reply CHANGELOG.md]: Done

**Comment by tbhaxor** [comment]
@pgrete What do you think?

**Review by pgrete — APPROVED** [review]
Thanks for the changes. Looks good now.

**Comment by tbhaxor** [comment]
Glad to become contributor of parthenon :)

**Comment by Yurlungur** [comment]
Enabling auto merge. Final tests should run now.

**Comment by tbhaxor** [comment]
@Yurlungur Does **Check Python and C++ Formatting** usually takes a lot of time?

**Comment by Yurlungur** [comment]
@tbhaxor no that's unusual. Let me see what's going on

**Comment by Yurlungur** [comment]
@tbhaxor it may be because you're on a fork. Can you go into your repository settings and look under actions/general and see what the settings are? You may need to whitelist the following:
```C++
gregsdennis/dependencies-action@*,
peaceiris/actions-gh-pages@*,
pgrete/cpp-py-formatter/*@*,
```

**Comment by tbhaxor** [comment]
@Yurlungur Done

Screenshot_20230320_121255

**Comment by pgrete** [comment]
@par-hermes format

**Comment by tbhaxor** [comment]
@par-hermes format

**Comment by pgrete** [comment]
Could you please run the `format` build target on your branch to see if the style is correct?
It looks like our formatting "service" does not work on forks.

**Comment by tbhaxor** [comment]
I checked there is no formatting issue. You can force merge this PR.

Regards,
Gurkirat Singh
(Website: https://tbhaxor.com)

**Comment by tbhaxor** [comment]
@Yurlungur @pgrete do you want to me push directly to the parthenon-hpc-lab/parthenon repository in future if I had some improvements or via fork?

**Pull request merged**
